### ISSUE TITLE
What reason will cause 'Provided key couldn't be added to the keychain' error

### ISSUE THREAD

**Issue by heqichang** [TakeScoop/SwiftyRSA#27]
Hi,

I'm not familiar with RSA, would you explain why my code will cause 'Provided key couldn't be added to the keychain' once in a while.

```swift
func encryptStringUseRsa(str:String)->String?{
        let bundle = NSBundle.mainBundle()
        
        let pubPath   = bundle.pathForResource("rsa_public_key", ofType: "pem")!
        let pubString = (try! NSString(contentsOfFile: pubPath, encoding: NSUTF8StringEncoding)) as String
        let encrypted = try! SwiftyRSA.encryptString(str, publicKeyPEM: pubString,padding: SecPadding.PKCS1)

        return encrypted
}
```

**Comment by heqichang** [comment]
Hi,

May be I found a solution to fix my error. I changed my code as below: 

```swift
func encryptStringUseRsa(str:String)->String?{
        let rsa = SwiftyRSA()
        if BHTokenCreator.publicKey == nil {
            let bundle = NSBundle.mainBundle()
            let pubPath   = bundle.pathForResource("rsa_public_key", ofType: "pem")!
            let pubString = (try! NSString(contentsOfFile: pubPath, encoding: NSUTF8StringEncoding)) as String
            BHTokenCreator.publicKey = try! rsa.publicKeyFromPEMString(pubString)
        }
        
        let encrypted = try! rsa.encryptString(str, publicKey: BHTokenCreator.publicKey!)

        return encrypted
}
```

There are no need add key to keychain every time. (?)
But still I don't know why it will cause error.

**Comment by ldiqual** [comment]
@heqichang Thanks for opening an issue! Public/private keys are persisted to your keychain until the `SwiftyRSA` instance gets deallocated. You cannot add the same key twice to the same instance of `SwiftyRSA`. However it doesn't seem to be the case here, because `rsa` should get deallocated at the end of the `encryptStringUseRsa` function. Are you sure you're not storing `rsa` in a bigger scope – like a view controller?

**Comment by heqichang** [comment]
@ldiqual Thank you for you reply. I'm sure I used `SwiftyRSA` only in the `encryptStringUseRsa` function. But I called this method in my every api method that may be called by multiple thread at same time.
I set breakpoint at below code in `SwiftRSA`:
```swift
var secStatus = SecItemAdd(keyDict as CFDictionary, persistKey)
if secStatus != noErr && secStatus != errSecDuplicateItem {
     throw SwiftyRSAError(message: "Provided key couldn't be added to the keychain")
}
```
secStatus got error -34018.

**Comment by heqichang** [comment]
Hi,

Still have same problem. Seems like this is a bug in Keychain. [http://stackoverflow.com/questions/29740952/osstatus-error-code-34018](url)

**Comment by ldiqual** [comment]
@heqichang Thanks for following up. I'll look into it and get back to you soon. Thanks!

**Comment by ldiqual** [comment]
@heqichang I overlooked your last comment and indeed it seems like a bug in the keychain. This issue will serve as future reference for people who encounter the same status code. Thank you!

**Comment by rchatham** [comment]
Are there any solutions for this? I am getting the equivalent "Couldn't create key reference from key data" in iOS10.

**Comment by ldiqual** [comment]
@rchatham This is fixed in the latest version of SwiftyRSA (0.5.0). Check it out on cocoapods! Sorry for the confusion, I should have referenced the commit here. https://github.com/TakeScoop/SwiftyRSA/commit/05dc93d16b0ea8da84125c5d4a2fa2e77403d58f#diff-b0f6a9136332456435064f29ad23f0d7R563
